### PR TITLE
codegen_llvm: check inline assembly constraints with LLVM

### PR DESCRIPTION
**Pull request by levex**
---%<---
Hey all,

As issue #54130 highlights, constraints are not checked and passing bad constraints to LLVM can crash it since a `Verify()` call is placed inside an assertion (see: `src/llvm/lib/IR/InlineAsm.cpp:39`).

As this is my first PR to the Rust compiler (woot! :tada:), there might be better ways of achieving this result. In particular, I am not too happy about generating an error in codegen; it would be much nicer if we did it earlier. However, @rkruppe [noted on IRC](https://botbot.me/mozilla/rustc/2018-09-25/?msg=104791581&page=1) that this should be fine for an unstable feature and a much better solution than the _status quo_, which is an ICE.

Thanks!
--->%---

LLVM provides a way of checking whether the constraints and the actual
inline assembly make sense. This commit introduces a check before
emitting code for the inline assembly. If LLVM rejects the inline
assembly (or its constraints), then the compiler emits an error E0668
("malformed inline assembly").

Fixes: #54130
Signed-off-by: Levente Kurusa \<lkurusa@acm.org\>

### CODE REVIEW

File: src/librustc_codegen_llvm/builder.rs [hunk 1]
```diff
@@ -753,9 +753,17 @@ impl Builder<'a, 'll, 'tcx> {
         debug!("Asm Output Type: {:?}", output);
         let fty = Type::func(&argtys[..], output);
         unsafe {
-            let v = llvm::LLVMRustInlineAsm(
-                fty, asm, cons, volatile, alignstack, dia);
-            self.call(v, inputs, None)
+            // Ask LLVM to verify that the constraints are well-formed.
+            let constraints_ok = llvm::LLVMRustInlineAsmVerify(fty, cons);
+            debug!("Constraint verification result: {:?}", constraints_ok);
+            if constraints_ok == 1 {
```
Comment (nagisa) [inline src/librustc_codegen_llvm/builder.rs]: Compare with `llvm::True` instead.

Comment (levex) [reply src/librustc_codegen_llvm/builder.rs]: Fixed in 70bf903 !

**Comment by levex** [comment]
Travis passed finally :tada: !

**Comment by nagisa** [comment]
@levex please also add a few tests which would specifically test the functionality introduced by this PR.

These tests would most likely end up going to `src/test/ui/`.

**Pull request merged**